### PR TITLE
atomicAdd for previous versions of CUDA [atmcadd]

### PR DESCRIPTION
**Pull request by camierjs**
Adds an implementation of `atomicAdd` for `CUDA_ARCH` before `600`. 

Adds by the way an `OpenMP` fix of the `AtomicAdd`.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#2475](https://github.com/mfem/mfem/pull/2475) | @camierjs | @tzanio | @YohannDudouit + @artv3 | 08/18/21 | 09/02/21 | 09/05/21 | |

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: general/backends.hpp [hunk 1]
```diff
@@ -66,6 +83,9 @@ MFEM_HOST_DEVICE T AtomicAdd(T &add, const T val)
    return atomicAdd(&add,val);
 #else
    T old = add;
+#ifdef MFEM_USE_OPENMP
+   #pragma omp atomic
+#endif
```
Comment (artv3) [inline general/backends.hpp:88]: Yes!!!!

**Review by artv3 — APPROVED** [review]

**Comment by tzanio** [comment]
@YohannDudouit this should be an easy review, can you take a quick look?

**Review by YohannDudouit — APPROVED** [review]
Compilation works on `sm_52` (tested on tux402), but I get compilation errors when trying to run the tests (not sure this is linked to this PR):
```
/home/dudouit1/Projects/mfem/tests/unit/general/test_text.cpp(18): warning: variable "<unnamed>::autoRegistrar1" was declared but never referenced

/home/dudouit1/Projects/mfem/tests/unit/general/test_array.cpp: In function ‘void ____C_A_T_C_H____T_E_S_T____0()’:
/home/dudouit1/Projects/mfem/tests/unit/general/test_array.cpp:21:64: error: no matching function for call to ‘mfem::Array<int>::Array(<brace-enclosed initializer list>)’
    Array<int> b({6.0, 5.0, 4.0, 3.0, 2.0, 1.0});
                                                                ^
/home/dudouit1/Projects/mfem/tests/unit/general/test_array.cpp:21:64: note: candidates are:
../../general/array.hpp:88:44: note: template<class CT, int N> mfem::Array<T>::Array(const CT (&)[N])
    template <typename CT, int N>
                                            ^
../../general/array.hpp:88:44: note:   template argument deduction/substitution failed:
/home/dudouit1/Projects/mfem/tests/unit/general/test_array.cpp:21:64: note:   couldn't deduce template parameter ‘CT’
    Array<int> b({6.0, 5.0, 4.0, 3.0, 2.0, 1.0});
                                                                ^
../../general/array.hpp:84:28: note: template<class CT> mfem::Array<T>::Array(const mfem::Array<CT>&)
    template <typename CT>
                            ^
../../general/array.hpp:84:28: note:   template argument deduction/substitution failed:
/home/dudouit1/Projects/mfem/tests/unit/general/test_array.cpp:21:64: note:   couldn't deduce template parameter ‘CT’
    Array<int> b({6.0, 5.0, 4.0, 3.0, 2.0, 1.0});
                                                                ^
../../general/array.hpp:632:1: note: mfem::Array<T>::Array(const mfem::Array<T>&) [with T = int]
 inline Array<T>::Array(const Array &src)
 ^
../../general/array.hpp:632:1: note:   no known conversion for argument 1 from ‘<brace-enclosed initializer list>’ to ‘const mfem::Array<int>&’
../../general/array.hpp:76:1: note: mfem::Array<T>::Array(T*, int) [with T = int]
    inline Array(T *data_, int asize)
 ^
../../general/array.hpp:76:1: note:   candidate expects 2 arguments, 1 provided
../../general/array.hpp:70:10: note: mfem::Array<T>::Array(int) [with T = int]
    explicit inline Array(int asize)
          ^
../../general/array.hpp:70:10: note:   no known conversion for argument 1 from ‘<brace-enclosed initializer list>’ to ‘int’
../../general/array.hpp:67:1: note: mfem::Array<T>::Array(mfem::MemoryType) [with T = int]
    inline Array(MemoryType mt) : size(0) { data.Reset(mt); }
 ^
../../general/array.hpp:67:1: note:   no known conversion for argument 1 from ‘<brace-enclosed initializer list>’ to ‘mfem::MemoryType’
../../general/array.hpp:64:9: note: mfem::Array<T>::Array() [with T = int]
    inline Array() : size(0) { data.Reset(); }
         ^
../../general/array.hpp:64:9: note:   candidate expects 0 arguments, 1 provided
make[2]: *** [general/test_array.o] Error 1
```

**Comment by v-dobrev** [comment]
@YohannDudouit, I think I've seen similar errors with gcc 4.8.x (maybe it does not fully support c++11?) -- try using gcc >= 4.9.x.

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Review by tzanio — APPROVED** [review]
Thanks for adding this John!

**Pull request merged**